### ISSUE TITLE
Appending html to div breaks canvas inside div.

### ISSUE THREAD

**Issue by Tigpan-Bytes** [processing/p5.js#4064]
#### Nature of issue?

- [X] Found a bug
- [ ] Existing feature enhancement
- [ ] New feature request

#### Most appropriate sub-area of p5.js?

- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [X] Other (specify if possible) (DOM)

#### Which platform were you using when you encountered this?

- [ ] Mobile/Tablet (touch devices)
- [X] Desktop/Laptop
- [ ] Others (specify if possible)

#### Details about the bug: 

- p5.js version: 0.9.0
- Web browser and version: Tested on latest verisions of Chrome and Firefox
- Operating System: Windows
- Steps to reproduce this: https://codepen.io/tigpan-bytes/pen/mddymmN -> Then press g to watch the bug

I was creating a chat room using socket.io and p5js when I was attempting to make pictures and drawings as something you can send. I then discovered that placing a canvas inside of a div element, then appending html to that div breaks the canvas permanently and irreversibly.

**Comment by lmccart** [comment]
hi @Tigpan-Bytes. the [`html()`](https://p5js.org/reference/#/p5.Element/child) method actually replaces the entire html contents, that's why you're seeing everything disappear. instead, you can use [`child()`](https://p5js.org/reference/#/p5.Element/child) which appends a child node. something like this:
```js
      let newText = p.createDiv('New line!\n');
      holder.child(newText);
```

**Comment by Tigpan-Bytes** [comment]
Strange, the documentation says differently. "If true is included as a second argument, html is appended instead of replacing existing html." If what you say is true, it sounds like the html method is very poorly optimized.

**Comment by lmccart** [comment]
Oh I see, I forgot that was an option for `html()`. This does seem like it could be a bug then.

**Comment by limzykenneth** [comment]
It is more or less working as expected. In the source [here](https://github.com/processing/p5.js/blob/master/src/dom/dom.js#L1635), the new content is appended with `this.elt.innerHTML += arguments[0];`, `+=` is an assignment and as such means that it will replace the content of `innerHTML`. When that content is replaced (even with the original content), everything is reset in it, in particular the canvas.

To get around this, we can replace it with [`append`](https://developer.mozilla.org/en-US/docs/Web/API/ParentNode/append).

**Comment by haschdl** [comment]
A bit long answer. 

### Part 1: About the example
> you can use [`child()`](https://p5js.org/reference/#/p5.Element/child) which appends a child node. something like this:
> 
> ```js
>       let newText = p.createDiv('New line!\n');
>       holder.child(newText);
> ```

I agree with the above. In fact, I would also recommend a different approach, which might prove more useful "down the road" when you want to apply styles. Instead of writing HTML (`"New line!\n"`), you could create a new `p` element (p is for paragraph), using createP.  

So instead of doing:
```js
if (p.keyCode == 71) // g
    { // appends new html to the div when g is pressed
        holder.html("New line!\n", true);
    }
```
You could do...
```js
  if (p.keyCode == 71) // g
    { 
      // appends new html to the div when g is pressed
      let pElt = p.createP("New 'line' (as a paragraph element)");
      holder.child(pElt);
    }
```
### Part 2: The bug

As for the bug... I believe the implementation of [html()](https://p5js.org/reference/#/p5.Element/html) with the second argument should not use the `innerHTML`, because it has the side-effect of forcing a reparse of the DOM - I can only guess that doing that breaks the "animation loop", and the result is that the sketch disappears. 

A better solution, which I tested in the codepen provided, is to use [insertAdjacentHTML](https://developer.mozilla.org/en-US/docs/Web/API/Element/insertAdjacentHTML). 

```js
if (p.keyCode == 71) // g
    { 
      // appends new html to the div when g is pressed
      holder.elt.insertAdjacentHTML("beforeend", "New line!"); 
    }
```
![image](https://user-images.githubusercontent.com/9550197/66810391-795ab580-ef2f-11e9-8ff8-29f266a5124e.png)

In other words, p5.js `html()` function could be: 
```js
p5.Element.prototype.html = function() {
  if (arguments.length === 0) {
    return this.elt.innerHTML;
  } else if (arguments[1]) {
    this.elt.insertAdjacentHTML('beforeend', arguments[0]);
    return this;
  } else {
    this.elt.innerHTML = arguments[0];
    return this;
  }
};
```

**I would propose to check if @Tigpan-Bytes agrees with the workaround provided, and we either create a new issue with a more focused description of the side-effect of changing innerHTML in the `child()` funtion**. How does that sound?

**Comment by limzykenneth** [comment]
`insertAdjacentHTML` may be a better choice as it parses the DOM string while `append` will need the object passed to be a node object or it will just parse as pure text.

**Comment by Tigpan-Bytes** [comment]
In response to @haschdl, thanks for the explanation and I would agree that creating a new issue would be the best course of action. Feel free to create that issue if you deem it necessary.

**Comment by haschdl** [comment]
Hello! After reconsidering the magnitude of the change I think another issue is overkill. I do however want to post a minimal example, in case the Codepen example goes offline.

```javascript
 let canvas;
 let canvasContainer;

 function setup() {
   canvasContainer = createDiv("Canvas will be added to this container");
   canvas = createCanvas(200, 200);
   canvas.parent(canvasContainer);
 }

 function draw() {
   background(0);
   circle(100, 100, 100); 
   
   
   //Calling html('something', 'true') on the 
   //div containing the sketch will breaks the animation
   canvasContainer.html("New line with p5.js html() function", true);
   
   //Calling insertAdjacentHTML (Javascript native) works
   //canvasContainer.elt.insertAdjacentHTML('beforeend', 'New line with appendAjacentHTML');
   
   noLoop();
 }
```
